http/README.md

```diff
@@ -0,0 +1,159 @@
+
+### Service Names
+
+Traditionally, libp2p protocols have used path-like protocol identifiers, e.g. `/libp2p/autonat/1.0.0`. Due to the use of `/`s, this doesn’t work well with the naming convention defined above.
+
+Protocols that wish to use the libp2p request-response mechanism MUST define a service name that is a valid URI component (according to RFC 8820).
+
+In practice, this isn’t expected cause too much friction, since current libp2p protocols were not designed to use the request-reponse mechanism, and will need to make arrangements to support it anyway (e.g. define how requests and responses are serialized).
+
+### Privacy Properties
+
+This leads to some very desirable properties:
+
+1. It is possible to run libp2p alongside a normal HTTP web service, i.e. on the same domain and port, without having to worry about collisions. 
+    1. As an on-path observer only sees SNI and ALPN, this effectively hides the fact that a client is establishing a connection in order to speak libp2p.
+2. Since authentication is flexible (see below), this enables servers to
+    1. require authentication to (some) paths below `.well-known/libp2p`, and to enforce ACLs
+    2. stealth mode: return 404 for paths below `.well-known/libp2p`, *unless* the client has already authenticated itself, thereby hiding the fact that it runs a libp2p server, even if probed explicitly
+
+## Certificates
+
+libp2p doesn’t prescribe how nodes obtain the TLS certificate to secure the HTTPS connection. Since browsers are expected to connect to the node, the certificate’s trust chain must end in the browser’s trust store.
+
+This is somewhat tricky in a p2p context, as nodes might not have a (sub)domain, which for many CAs is a requirement to obtain a certificate. Specifically, Let’s Encrypt doesn’t support IP certificates at the moment. ZeroSSL does, however, this requires setting up a (free) account.
+
+To speed of server authentication, a node MAY include the libp2p TLS extension in its certificate. Note that this is currently not possible when using Let’s Encrypt, since the libp2p TLS extension is not whitelisted by LE. Not every HTTP client will have access to the TLS certificate (for example, browsers usually don’t expose an API for that), but if an HTTP client does, it SHOULD use that information.
+
+## Authentication
+
+Traditionally, libp2p was built on the assumption that both peers authenticate each other during the libp2p handshake. libp2p+HTTP acknowledges that this isn’t always possible, or even desirable, and that different use cases call for different authentication modes. For example, a server might offer a certain set of services to any client, like a HTTP webserver does.
+
+### Server Authentication
+
+Since HTTP requests are independent from each other (they are not bound to a single connection, and when using HTTP/1.1, will actually use different connections), the server needs to authenticate itself on every single request.
```

Comment:
Maybe? On the other hand, if you're willing to do a TLS handshake for every HTTP/1.1 request, maybe you don't really care about a signature. Of course this will look different for a HTTP/2 or HTTP/3 connection...
When caching, you'd have to be very careful about HTTP redirects.

There's also the risk that this signature scheme negates the benefits of using a CDN: Unless the libp2p key is at the edge, every request has to go back to the origin to perform this signature.

Reply:
> these things together mean the responses for the same thing sent to different users will have different header, and will not be reusable across users. Effectively making it impossible to benefit from HTTP caching across users.

Is that so? HTTP headers shouldn't have any effect on cacheability, unless you include them in the [Vary](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Vary) header.

> For example, DHT lookup for a peer records would hit the backend server every time, even tho you could safely cache results for a while at the edge CDN.

The first question I'd ask is if it makes sense to request authentication in that case. I imagine that _almost all_ requests we handle won't need peer authentication, since we're dealing with either signed or content-addressed data anyway.